### ISSUE TITLE
BUG: overlay union raises ValueError with pandas 1.5.0rc0

### ISSUE THREAD

**Issue by zaneselvans** [geopandas/geopandas#2544]
- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the latest version of geopandas.
- [ ] (optional) I have confirmed this bug exists on the main branch of geopandas.

---

### Code Sample, a copy-pastable example
```python
from geopandas import GeoDataFrame, GeoSeries, overlay, show_versions
from shapely.geometry import Polygon

gdf_a = GeoDataFrame({"geometry": GeoSeries([Polygon([(0, 0), (0, 1), (3, 1), (3, 0)])])})
gdf_b = GeoDataFrame({"geometry": GeoSeries([Polygon([(2, 0), (2, 1), (4, 1), (4, 0)])])})

overlay(gdf_a, gdf_b, how="union")
```

### Problem description

While trying to get [some of our code](https://github.com/catalyst-cooperative/pudl/issues/1901) to work with the pandas 1.5.0rc0, I found that the above code raises:
```
ValueError: GeoDataFrame does not support multiple columns using the geometry column name 'geometry'.
```

Traceback here:
<details>

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [1], in <cell line: 11>()
      4 gdf_a = GeoDataFrame(
      5     {"geometry": GeoSeries([Polygon([(0, 0), (0, 1), (3, 1), (3, 0)])])}
      6 )
      7 gdf_b = GeoDataFrame(
      8     {"geometry": GeoSeries([Polygon([(2, 0), (2, 1), (4, 1), (4, 0)])])}
      9 )
---> 11 overlay(gdf_a, gdf_b, how="union")

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/geopandas/tools/overlay.py:321, in overlay(df1, df2, how, keep_geom_type, make_valid)
    319     result = _overlay_symmetric_diff(df1, df2)
    320 elif how == "union":
--> 321     result = _overlay_union(df1, df2)
    322 elif how == "identity":
    323     dfunion = _overlay_union(df1, df2)

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/geopandas/tools/overlay.py:137, in _overlay_union(df1, df2)
    133 """
    134 Overlay Union operation used in overlay function
    135 """
    136 dfinter = _overlay_intersection(df1, df2)
--> 137 dfsym = _overlay_symmetric_diff(df1, df2)
    138 dfunion = pd.concat([dfinter, dfsym], ignore_index=True, sort=False)
    139 # keep geometry column last

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/geopandas/tools/overlay.py:117, in _overlay_symmetric_diff(df1, df2)
    115 _ensure_geometry_column(dfdiff2)
    116 # combine both 'difference' dataframes
--> 117 dfsym = dfdiff1.merge(
    118     dfdiff2, on=["__idx1", "__idx2"], how="outer", suffixes=("_1", "_2")
    119 )
    120 geometry = dfsym.geometry_1.copy()
    121 geometry.name = "geometry"

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/geopandas/geodataframe.py:1499, in GeoDataFrame.merge(self, *args, **kwargs)
   1478 def merge(self, *args, **kwargs):
   1479     r"""Merge two ``GeoDataFrame`` objects with a database-style join.
   1480 
   1481     Returns a ``GeoDataFrame`` if a geometry column is present; otherwise,
   (...)
   1497 
   1498     """
-> 1499     result = DataFrame.merge(self, *args, **kwargs)
   1500     geo_col = self._geometry_column_name
   1501     if isinstance(result, DataFrame) and geo_col in result:

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/pandas/core/frame.py:10111, in DataFrame.merge(self, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)
  10092 @Substitution("")
  10093 @Appender(_merge_doc, indents=2)
  10094 def merge(
   (...)
  10107     validate: str | None = None,
  10108 ) -> DataFrame:
  10109     from pandas.core.reshape.merge import merge
> 10111     return merge(
  10112         self,
  10113         right,
  10114         how=how,
  10115         on=on,
  10116         left_on=left_on,
  10117         right_on=right_on,
  10118         left_index=left_index,
  10119         right_index=right_index,
  10120         sort=sort,
  10121         suffixes=suffixes,
  10122         copy=copy,
  10123         indicator=indicator,
  10124         validate=validate,
  10125     )

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/pandas/core/reshape/merge.py:125, in merge(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)
     94 @Substitution("\nleft : DataFrame or named Series")
     95 @Appender(_merge_doc, indents=0)
     96 def merge(
   (...)
    109     validate: str | None = None,
    110 ) -> DataFrame:
    111     op = _MergeOperation(
    112         left,
    113         right,
   (...)
    123         validate=validate,
    124     )
--> 125     return op.get_result(copy=copy)

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/pandas/core/reshape/merge.py:777, in _MergeOperation.get_result(self, copy)
    773     self.left, self.right = self._indicator_pre_merge(self.left, self.right)
    775 join_index, left_indexer, right_indexer = self._get_join_info()
--> 777 result = self._reindex_and_concat(
    778     join_index, left_indexer, right_indexer, copy=copy
    779 )
    780 result = result.__finalize__(self, method=self._merge_type)
    782 if self.indicator:

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/pandas/core/reshape/merge.py:767, in _MergeOperation._reindex_and_concat(self, join_index, left_indexer, right_indexer, copy)
    763 right.index = join_index
    765 from pandas import concat
--> 767 result = concat([left, right], axis=1, copy=copy)
    768 result.columns = llabels.append(rlabels)
    769 return result

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/pandas/util/_decorators.py:317, in deprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
    311 if len(args) > num_allow_args:
    312     warnings.warn(
    313         msg.format(arguments=arguments),
    314         FutureWarning,
    315         stacklevel=find_stack_level(inspect.currentframe()),
    316     )
--> 317 return func(*args, **kwargs)

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/pandas/core/reshape/concat.py:382, in concat(objs, axis, join, ignore_index, keys, levels, names, verify_integrity, sort, copy)
    160 """
    161 Concatenate pandas objects along a particular axis.
    162 
   (...)
    367 1   3   4
    368 """
    369 op = _Concatenator(
    370     objs,
    371     axis=axis,
   (...)
    379     sort=sort,
    380 )
--> 382 return op.get_result()

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/pandas/core/reshape/concat.py:624, in _Concatenator.get_result(self)
    621     new_data._consolidate_inplace()
    623 cons = sample._constructor
--> 624 return cons(new_data).__finalize__(self, method="concat")

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/geopandas/geodataframe.py:32, in _geodataframe_constructor_with_fallback(*args, **kwargs)
     26 def _geodataframe_constructor_with_fallback(*args, **kwargs):
     27     """
     28     A flexible constructor for GeoDataFrame._constructor, which falls back
     29     to returning a DataFrame (if a certain operation does not preserve the
     30     geometry column)
     31     """
---> 32     df = GeoDataFrame(*args, **kwargs)
     33     geometry_cols_mask = df.dtypes == "geometry"
     34     if len(geometry_cols_mask) == 0 or geometry_cols_mask.sum() == 0:

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/geopandas/geodataframe.py:160, in GeoDataFrame.__init__(self, data, geometry, crs, *args, **kwargs)
    151 if (
    152     geometry is None
    153     and self.columns.nlevels == 1
   (...)
    157     # self["geometry"] is a gdf and constructor gets recursively recalled
    158     # by pandas internals trying to access this
    159     if (self.columns == "geometry").sum() > 1:
--> 160         raise ValueError(
    161             "GeoDataFrame does not support multiple columns "
    162             "using the geometry column name 'geometry'."
    163         )
    165     # only if we have actual geometry values -> call set_geometry
    166     try:

ValueError: GeoDataFrame does not support multiple columns using the geometry column name 'geometry'.
```

</details>

### Expected output / behavior

The code snippet works as expected and returns the following geodataframe when using pandas 1.4.4:

|    | geometry                            |
|---:|:------------------------------------|
|  0 | POLYGON ((3 1, 3 0, 2 0, 2 1, 3 1)) |
|  1 | POLYGON ((0 1, 2 1, 2 0, 0 0, 0 1)) |
|  2 | POLYGON ((4 1, 4 0, 3 0, 3 1, 4 1)) |

### Output of ``geopandas.show_versions()``

<details>

SYSTEM INFO
-----------
python     : 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:35:26) [GCC 10.4.0]
executable : /home/zane/mambaforge/envs/pudl-dev/bin/python
machine    : Linux-5.15.0-47-generic-x86_64-with-glibc2.35

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.11.0
GEOS lib   : /home/zane/mambaforge/envs/pudl-dev/lib/libgeos_c.so
GDAL       : 3.5.1
GDAL data dir: /home/zane/mambaforge/envs/pudl-dev/share/gdal
PROJ       : 9.0.1
PROJ data dir: /home/zane/mambaforge/envs/pudl-dev/share/proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.11.1
numpy      : 1.22.4
pandas     : 1.5.0rc0
pyproj     : 3.3.1
shapely    : 1.8.4
fiona      : 1.8.21
geoalchemy2: None
geopy      : None
matplotlib : 3.5.3
mapclassify: 2.4.3
pygeos     : 0.12.0
pyogrio    : None
psycopg2   : None
pyarrow    : 9.0.0
rtree      : 1.0.0

</details>

**Comment by m-richards** [comment]
Hi @zaneselvans thanks for reporting this, seems to be the same issue as what's causing issues on the CI for main.

Looking a bit deeper, this is caused by https://github.com/pandas-dev/pandas/pull/48082 and it seems @jorisvandenbossche had already noted it. 

It seems like this could be fixed in geopandas, but to do so we'd have to avoid using suffixes in `pd.merge` in `overlay` and `sjoin` and instead manually re-key the collisions in column names. This isn't ideal though since we'd then be duplicating / reimplementing the name collision logic from pandas itself, so an upstream fix would be better.

I guess it's worth noting that this tension between our `_constructor` and `__finalize__` checks and column names only being applied after an internal pandas operation has come up before in a couple of other PRs.

**Comment by jorisvandenbossche** [comment]
Yes, I was planning to fix this on the pandas side, but only opened a PR now with some delay (https://github.com/pandas-dev/pandas/pull/48416)

In general I would certainly prefer that we handle this in pandas (although it's a patch just for us), because while we can workaround it in the sjoin code, that also means that any end-user doing a merge with suffices would have to do the same workaround of renaming manually.

> I guess it's worth noting that this tension between our `_constructor` and `__finalize__` checks and column names only being applied after an internal pandas operation has come up before in a couple of other PRs.

Do you remember in which other cases this has come up?

**Comment by m-richards** [comment]
> Do you remember in which other cases this has come up?

The only case I could find was this one: https://github.com/geopandas/geopandas/pull/2296#issuecomment-1024841891 which is actually a bit different, it's about default column labels being updated with correct labels, rather than suffixes not being applied. I don't think is actually an issue any more in this particular case, as that PR eventually did not change the behaviour as described - and the discussion in https://github.com/geopandas/geopandas/issues/2364 would I think suggest the geometry column would be unset.

I think there may have been another case, where there was something I couldn't quite work out because of this but I can't seem to find the PR discussing it, which is not super useful.

**Comment by jorisvandenbossche** [comment]
This was fixed on the pandas side before the final 1.5, so closing.
